src/index.rs

```diff
@@ -182,17 +180,15 @@ impl Index {
                     )
                 }
                 let blockhashes: Vec<BlockHash> = chunk.iter().map(|h| h.hash()).collect();
-                let mut heights_map: HashMap<BlockHash, usize> =
-                    chunk.iter().map(|h| (h.hash(), h.height())).collect();
+                let mut heights = chunk.iter().map(|h| h.height());
 
                 let mut batch = WriteBatch::default();
-
-                daemon.for_blocks(blockhashes, |blockhash, block| {
-                    let height = heights_map.remove(&blockhash).expect("unexpected block");
+                daemon.for_blocks(blockhashes, |_blockhash, block| {
+                    let height = heights.next().expect("unexpected block");
                     let result = index_single_block(block, height);
                     result.extend(&mut batch);
                 })?;
-                assert!(heights_map.is_empty(), "some blocks were not indexed");
+                assert_eq!(heights.next(), None, "some blocks were not indexed");
```

Comment:
Since presumably you use `assert_eq!` instead of `is_none()` to get debug output containing the hash of not-indexed block maybe use `heights.collect::<Vec<_>>()` and compare to `Vec::new()`? That way we have no allocation in happy path and learn all extra hashes in debug output if things go wrong.

Or, if you want to get rid of allocations completely, use `chunks_mut` and `drain(..)` instead of `iter()`, then just check if chunk is empty at the end. :)

Reply:
Sounds good, changed to `collect`.